beets/metadata_plugins.py

```diff
@@ -357,3 +372,83 @@ def _construct_search_query(
             query = unidecode.unidecode(query)
 
         return query
+
+
+# To have proper typing for the proxy class below, we need to
+# trick mypy into thinking that SafeProxy is a subclass of
+# MetadataSourcePlugin.
+# https://stackoverflow.com/questions/71365594/how-to-make-a-proxy-object-with-typing-as-underlying-object-in-python
+Proxied = TypeVar("Proxied", bound=MetadataSourcePlugin)
+if TYPE_CHECKING:
+    base = MetadataSourcePlugin
+else:
+    base = object
+
+
+class SafeProxy(base):
+    """A proxy class that forwards all attribute access to the wrapped
+    MetadataSourcePlugin instance.
+
+    We use this to catch and log exceptions from metadata source plugins
+    without crashing beets. E.g. on long running autotag operations.
+    """
+
+    __plugin: MetadataSourcePlugin
+
+    def __init__(self, plugin: MetadataSourcePlugin):
+        self.__plugin = plugin
+
+    def __getattribute__(self, name):
+        if name in {
+            "_SafeProxy__plugin",
+            "_SafeProxy__handle_exception",
+            "candidates",
+            "item_candidates",
+            "album_for_id",
+            "track_for_id",
+        }:
+            return super().__getattribute__(name)
+        else:
+            return getattr(self.__plugin, name)
+
+    def __setattr__(self, name, value):
+        if name == "_SafeProxy__plugin":
+            super().__setattr__(name, value)
+        else:
+            self.__plugin.__setattr__(name, value)
+
+    def __handle_exception(self, func: Callable[P, R], e: Exception) -> None:
+        """Helper function to log exceptions from metadata source plugins."""
+        if config["raise_on_error"].get(bool):
+            raise e
```

Comment:
I think we would preferably handle this configuration option in `find_metadata_source_plugins`. This way, when we have, `raise_on_error: no`, exceptions will be raised _truly_ directly. This way, we also have a fallback in case something goes bad with `SafeProxy` implementation, since we have an option to disable it.

Reply:
Sounds reasonable. I was thinking we handle it here incase we want to attach the option to each plugin instead. Would allow for more granual control.

Although, I'm not sure we want to go this route.

Reply:
I cannot really see a use case for granular control here.

FYI `beetcamp` already handles its exceptions, where it points users to a URL to submit a GitHub issue when something unexpected happens.

Reply:
You are right, plugins should handle their exceptions gracefully! I think this logic here is just a last resort in-case something really unexpected happens.

I moved the config check into the `find_metadata_source_plugins` function.